kura/org.eclipse.kura.cloudconnection.sparkplug.mqtt.provider/src/main/java/org/eclipse/kura/cloudconnection/sparkplug/mqtt/endpoint/SparkplugCloudEndpoint.java

```diff
@@ -93,11 +108,50 @@ public void deactivate() {
      */
 
     @Override
-    public String publish(KuraMessage message) throws KuraException {
-        // TODO Auto-generated method stub
+    public String publish(final KuraMessage message) throws KuraException {
+        Map<String, Object> messageProperties = message.getProperties();
+        if (!messageProperties.containsKey(SparkplugDevice.KEY_MESSAGE_TYPE)
+                || !messageProperties.containsKey(SparkplugDevice.KEY_DEVICE_ID)) {
+            throw new KuraException(KuraErrorCode.INVALID_PARAMETER,
+                    "KuraMessage has a missing property between message.type and device.id");
+        }
+
+        SparkplugMessageType type = (SparkplugMessageType) messageProperties.get(SparkplugDevice.KEY_MESSAGE_TYPE);
+        String deviceId = (String) messageProperties.get(SparkplugDevice.KEY_DEVICE_ID);
+
+        logger.debug("{} - Sending message with seq: {}", this.kuraServicePid, this.seqCounter.getCurrent());
+
+        byte[] sparkplugPayload = SparkplugPayloads.getSparkplugDevicePayload(this.seqCounter.getCurrent(),
+                message.getPayload());
+
+        this.seqCounter.next();
+
+        if (type == SparkplugMessageType.DBIRTH) {
+            return publishInternal(
+                    SparkplugTopics.getDeviceBirthTopic(PLACEHOLDER_GROUP_ID, PLACEHOLDER_NODE_ID, deviceId),
+                    sparkplugPayload, 0, false, 0);
```

Comment:
qos is fixed to 0?

Reply:
Yes

[tck-id-message-flow-device-birth-publish-dbirth-qos] -> https://github.com/eclipse-sparkplug/sparkplug/blob/master/specification/src/main/asciidoc/chapters/Sparkplug_5_Operational_Behavior.adoc#device-session-establishment